readme.md

```diff
@@ -0,0 +1,167 @@
+
+## Description <a name="description"></a>
+
+This is an SDK for interacting with [CMS Blue Button 2.0 API](https://bluebutton.cms.gov/developers/),
+the API allows applications to obtain a beneficiary's (who has login account with medicare.gov) grant
+to access his/her medicare claims data - through OAUTH2 [(RFC 6749)](https://datatracker.ietf.org/doc/html/rfc6749) authorization flow.
+
+By using the SDK, the development of applications accessing Blue Button 2.0 API can be greatly simplified.
+
+Note, following the OAUTH2 best practices, OAUTH2 PKCE etension [(RFC 7636)](https://datatracker.ietf.org/doc/html/rfc7636) is always enabled.
+
+## Installation <a name="installation"></a>
+
+Using npm:
+
+```bash
+$ npm install cms-bluebutton-sdk
+```
+
+When develop with typescript
+
+```bash
+$ npm install --save-dev @types/cms-bluebutton-sdk
+```
+
+Using yarn:
+
+```bash
+$ yarn add cms-bluebutton-sdk
+```
+
+When develop with typescript
+
+```bash
+$ yarn add --dev @types/cms-bluebutton-sdk
+```
+
+## Configuration <a name="configuration"></a>
+
+the SDK needs to be properly configured to work, the parameters are:
+
+- the app's credentials - client id, client secret
+- the app's callback url
+- the version number of the API
+- the app's environment (web location where the app is registered)
+
+the configuration is in json format and stored in a local file, the default location
+is current working directory with file name: .bluebutton-config.json
+
+A sample configuration json:
+
+```
+{
+  "clientId": "foo",
+  "clientSecret": "bar",
+  "callbackUrl": "https://www.fake.com/",
+}
+
+```
+
+| parameter    | value                   | Comments                        |
+| ------------ | ----------------------- | ------------------------------- |
+| clientId     | "foo"                   | oauth2 client id of the app     |
+| clientSecret | "bar"                   | oauth2 client secret of the app |
+| callbackUrl  | "https://www.fake.com/" | oauth2 callback url of the app  |
+
+For application registration and client id and client secret, please refer to:
+[Blue Button 2.0 API Docs - Try the API](https://bluebutton.cms.gov/developers/#try-the-api)
+
+## Sample Usages: Obtain Access Grant, Probe Scope, and Access Data <a name="usages"></a>
+
+Below are psuedo code snippets showing SDK used with node express server.
+
+```
+
+import express, { Router, Request, Response } from 'express';
+
+const app = express();
+
+const bb = new BlueButton();
+const authData = bb.generateAuthData();
+
+// AuthorizationToken holds access grant info:
+// access token, expire in, expire at, token type, scope, refreh token, etc.
+// it is associated with current logged in user in real app,
+// check SDK js docs for more details.
+
+let authToken: AuthorizationToken;
+
+const router = Router();
+
+// start authorize flow: response with URL to redirect to Medicare.gov beneficiary login
+router.get('/authorize/authurl', bb.generateAuthorizeUrl);
```

Comment:
Just taking another look through this and noticing that some of this wont work as written

Reply:
this works but it do the redirect in the FE code:
    async function goAuthorize() {
        const authUrlResponse = await axios.get(`/api/authorize/authurl`, { params: settingsState });
        window.location.href = authUrlResponse.data || '/';
    }    

I agree, it is better to put redirect together in one place.